vm/src/cformat.rs

```diff
@@ -0,0 +1,357 @@
+/// Implementation of Printf-Style string formatting
+/// [https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting]
+use num_bigint::{BigInt, Sign};
+use num_traits::Signed;
+use std::cmp;
+use std::str::FromStr;
+use crate::format::{parse_number, parse_precision};
+
+#[derive(Debug, PartialEq)]
+pub enum CFormatErrorType {
+    UnmatchedKeyParentheses,
+    MissingModuloSign,
+    UnescapedModuloSignInLiteral,
+    UnsupportedFormatChar(char),
+    IncompleteFormat,
+    Unimplemented,
+}
+
+// also contains how many chars the parsing function consumed
+type ParsingError = (CFormatErrorType, usize);
+
+pub struct CFormatError {
+    pub typ: CFormatErrorType,
+    pub index: usize,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CFormatPreconversor {
+    Repr,
+    Str,
+    Ascii,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CFormatCase {
+    Lowercase,
+    Uppercase,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CNumberType {
+    Decimal,
+    Octal,
+    Hex(CFormatCase),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CFloatType {
+    Exponent(CFormatCase),
+    PointDecimal,
+    General(CFormatCase),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CFormatType {
+    Number(CNumberType),
+    Float(CFloatType),
+    Character,
+    String(CFormatPreconversor),
+}
+
+bitflags! {
+    pub struct CConversionFlags: u32 {
+        const ALTERNATE_FORM = 0b0000_0001;
+        const ZERO_PAD = 0b0000_0010;
+        const LEFT_ADJUST = 0b0000_0100;
+        const BLANK_SIGN = 0b0000_1000;
+        const SIGN_CHAR = 0b0001_0000;
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct CFormatSpec {
+    pub mapping_key: Option<String>,
+    pub flags: CConversionFlags,
+    pub min_field_width: Option<usize>,
+    pub precision: Option<usize>,
+    pub format_type: CFormatType,
+    pub format_char: char,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CFormatPart {
+    Literal(String),
+    Spec(CFormatSpec),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct CFormatString {
+    pub format_parts: Vec<(usize, CFormatPart)>,
+}
+
+impl CFormatString {
```

Comment:
This "class" contains mostly static methods. This impl block might not be required, and we can use simple functions for the parsing, right?

Reply:
Yep, thanks